### ISSUE TITLE
Cannot scan for music

### ISSUE THREAD

**Issue by infidelus** [owncloud/music#774]
I recently upgraded to Music 0.15 on Nextcloud 19.0.0 but noticed not all my music was showing.

I decided to reset the music collection, but when I then tried to run a scan from the GUI it just sat there at 0/xxxxx files (no matter how long I left it).

Today I ssh'd into the server and ran 'sudo -u www-data ./occ music:scan myuser' and this got the following output:

`Start scan for myuser
Found 11169 new music files
Scanning /myuser/files/Music/MP3/Artist/Album/12 - Some Song.mp3

In Mapper.php line 281:
                                                                                                                
  Did expect one result but found none when executing: query "SELECT                                            
  					`*PREFIX*music_genres`.`id`,                                                                             
  					`*PREFIX*music_genres`.`name`,                                                                           
  					`*PREFIX*music_genres`.`lower_name`,                                                                     
  					COUNT(`track`.`id`) AS `trackCount`,                                                                     
  					COUNT(DISTINCT(`track`.`album_id`)) AS `albumCount`,                                                     
  					COUNT(DISTINCT(`track`.`artist_id`)) AS `artistCount`                                                    
  				FROM `*PREFIX*music_tracks` `track`                                                                       
  				INNER JOIN `*PREFIX*music_genres`                                                                         
  				ON `track`.`genre_id` = `*PREFIX*music_genres`.`id`                                                       
  				WHERE `track`.`genre_id` IS NOT NULL AND `*PREFIX*music_genres`.`user_id` = ? AND `lower_name` = ?        
  				GROUP BY `*PREFIX*music_genres`.`id`, `*PREFIX*music_genres`.`name`, `*PREFIX*music_genres`.`lower_name`  
  				HAVING COUNT(`track`.`id`) > 0"; parameters Array                                                         
  (                                                                                                             
      [0] => myuser                                                                                               
      [1] => pop                                                                                                
  )                                                                                                             
  ; limit ""; offset ""                                                                                         
                                                                                                                

music:scan [--all] [--group GROUP] [--debug] [--clean-obsolete] [--rescan] [--] [<user_id>...]`

The file mentioned at the start is always the same one, but when I play it through 'Files' it plays perfectly, as it does from my PC, and the tags look fine (all music tagged from Musicbrainz or Discogs).

I've also tried resetting the database using the occ command but no matter what, I get the above error.

**Comment by Biktorgj** [comment]
As a workaround (since it looks like a bug), if you empty (TRUNCATE) the music_genres table it will rescan correctly.

**Comment by infidelus** [comment]
I've already tried sudo -u www-data ./occ music:reset-database --all, and the error persisted.  Does that not empty the database?

**Comment by Biktorgj** [comment]
From the test's I've done, it looks like that is exactly the problem.

Steps to reproduce:
1. Have some media already scanned in Music app.
2. In Music, go to Settings and click on Reset Music Collection
3. Try to rescan media, it will get stuck in 0 songs scanned.

Temporary fix: 
Connect to your database server (for example with phpmyadmin), and in the Nextcloud database instance, run the following SQL command:

`TRUNCATE TABLE `oc_music_genres` `

Then try to rescan media. It should work now.

Somewhat correct fix (haven't looked much into the code but looks like this is where it should go:

In **/db/maintenance.php**, 
in Function **resetDb**

`	public function resetDb($userId, $allUsers = false) {
		if ($userId && $allUsers) {
			throw new InvalidArgumentException('userId should be null if allUsers targeted');
		}
	
		$sqls = [
				'DELETE FROM `*PREFIX*music_tracks`',
				'DELETE FROM `*PREFIX*music_albums`',
				'DELETE FROM `*PREFIX*music_artists`',
				'DELETE FROM `*PREFIX*music_playlists`',
				'DELETE FROM `*PREFIX*music_cache`'
		];

		foreach ($sqls as $sql) {
			$params = [];
			if (!$allUsers) {
				$sql .=  ' WHERE `user_id` = ?';
				$params[] = $userId;
			}
			$this->db->executeUpdate($sql, $params);
		}

		if ($allUsers) {
			$this->logger->log("Erased music databases of all users", 'info');
		} else {
			$this->logger->log("Erased music database of user $userId", 'info');
		}
	}`

Add 'DELETE FROM `*PREFIX*music_genres`' in the **$sqls** array

**Comment by infidelus** [comment]
Thanks.  The TRUNCATE TABLE command allowed me to scan my music.

I did however notice a significant amount of errors while it was scanning, which has resulted in a number of albums not showing.  For example:

`Scanning /myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac

{"reqId":"kbqygtnZs6ULDctWDwWc","level":3,"time":"2020-07-12T11:35:18+01:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"fopen(/mnt/ncdata/myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac): failed to open stream: Permission denied at /var/www/nextcloud/lib/private/Files/Storage/Local.php#351","userAgent":"--","version":"19.0.0.12"}

{"reqId":"kbqygtnZs6ULDctWDScanning /myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac

{"reqId":"kbqygtnZs6ULDctWDwWc","level":3,"time":"2020-07-12T11:35:18+01:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"fopen(/mnt/ncdata/myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac): failed to open stream: Permission denied at /var/www/nextcloud/lib/private/Files/Storage/Local.php#351","userAgent":"--","version":"19.0.0.12"}

{"reqId":"kbqygtnZs6ULDctWDwWc","level":3,"time":"2020-07-12T11:35:18+01:00","remoteAddr":"","user":"--","app":"music","method":"","url":"--","message":"Failed to open file /myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac for metadata extraction","userAgent":"--","version":"19.0.0.12"}
wWc","level":3,"time":"2020-07-12T11:35:18+01:00","remoteAddr":"","user":"--","app":"music","method":"","url":"--","message":"Failed to open file /myuser/files/Music/FLAC/Various/Now That's What I Call Music! 102/disc 2/16 - Relax.flac for metadata extraction","userAgent":"--","version":"19.0.0.12"}`

I've checked the files play in various desktop music players and checked the metadata in Puddletag, which also looks fine to me.

**Comment by infidelus** [comment]
Ignore that error.  Turns out some of my folders had different permissions.  Not sure how, but I did have to restore an older backup of my music after a failure about a month ago.  I've reset the permissions on the entire folder and it looks like it's now scanned everything.

I did still see a couple of errors and they appear to be related to one specific album, which won't play in the Music App:

`{"reqId":"XKxVqp8oVyHXN4UrCXs8","level":3,"time":"2020-07-12T12:43:43+01:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"fread(): Length parameter must be greater than 0 at /var/www/nextcloud/apps/music/3rdparty/getID3/getid3/getid3.php#2063","userAgent":"--","version":"19.0.0.12"}'

Then right at the end:

'{"reqId":"XKxVqp8oVyHXN4UrCXs8","level":3,"time":"2020-07-12T12:44:29+01:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"chmod(): Operation not permitted at /var/www/nextcloud/lib/private/Log/File.php#86","userAgent":"--","version":"19.0.0.12"}`

Looking at the second error, is it trying to change permissions?  The music directory is read-only so if that's what it's trying to do, it will fail.

**Comment by paulijar** [comment]
@infidelus, @Biktorgj Thanks for the report. Looking at the code, it is clear that there are two interconnected bugs here:
1) As @Biktorgj noticed, the DB reset functionality does not reset the `*PREFIX*music_genres` table as it should.
2) Scanning breaks, if the scanned track has a genre, which is already present in the DB, but the DB doesn't yet contain any tracks with that genre. Obviously, the bug 1) now creates such a situation.

What puzzles me is that why this bug 2) did not show up on my system. After reading this issue, I even tested resetting the DB and rescanning a couple of times, and everything just worked. Then I did some manual testing on the DB with the mysql client, and after that, the bug started to show up on my system, too. Go figure 🤷‍♂️. Anyway, I try to fix this and make a hotfix release still today.

@infidelus 
> I did still see a couple of errors and they appear to be related to one specific album, which won't play in the Music App

Can you play/download these problem files in the Nextcloud Files view?

> Looking at the second error, is it trying to change permissions? The music directory is read-only so if that's what it's trying to do, it will fail.

This error comes from the Nextcloud server code related to file logging. It seems to check if the log file has exactly the intended set of permissions. If not, then it tries to change those permissions. But if that fails, then it just goes on. As this line has ended up to the log file, obviously the file logging works regardless of the file having unexpected permissions :D. The desired set of permissions is read from `config.php`, and it defaults to `0640` if not found from the config.

**Comment by infidelus** [comment]
Thanks for the detailed reply @paulijar 

> Can you play/download these problem files in the Nextcloud Files view?

No, it just sits there at 'loading'.

**Comment by paulijar** [comment]
Music v0.15.1 with fixes for the two genre-related issues causing the scanning problem was just released.

The last mentioned problem of not being able to play some files is out of scope for the Music app, as the files can't be accessed by the Files app, either. Music app can play only what the cloud can access.

**Comment by paulijar** [comment]
@infidelus 
>No, it just sits there at 'loading'.

Can you still find the file from your file system? And does it seem to be okay?

**Comment by infidelus** [comment]
The files are all in my Music folder and I can play them quite happily with any desktop player.  The permissions are all the same (see  https://github.com/owncloud/music/issues/774#issuecomment-657215577) above.

**Comment by infidelus** [comment]
Quick update on the error playing one of my albums.  I still couldn't find anything wrong with it.  It played everywhere I tried.  After looking at the Nextcloud logs however, I saw:

> [PHP] Error: fread(): Length parameter must be greater than 0 at /var/www/nextcloud/apps/music/3rdparty/getID3/getid3/getid3.php#2063
> 
> GET /apps/music/api/album/92633/cover

I checked the cover and it displayed perfectly well in all my desktop music apps, as well as in the Nextcloud Music app.

Just to rule it out though I extracted the cover with Puddletag then removed, saved and re-added the exact same cover for all the tracks.  Lo and behold the album now plays in the Music app.

Go figure ...
